zsh-module/src/zsh.rs

```diff
@@ -47,33 +48,57 @@ pub fn eval_simple(cmd: &str) -> MaybeError<InternalError> {
     std::os::unix::io::FromRawFd::from_raw_fd(zsys::SHIN)
 } */
 
-#[derive(Debug)]
-#[repr(u32)]
-pub enum SourceError {
-    NotFound,
-    InternalError(InternalError),
-}
+// #[derive(Debug)]
+// #[repr(u32)]
+// pub enum SourceError {
+//     NotFound,
+//     InternalError(Zerror),
+// }
 
-impl std::fmt::Display for SourceError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            Self::NotFound => write!(f, "File not found"),
-            Self::InternalError(e) => e.fmt(f),
-        }
-    }
-}
-impl std::error::Error for SourceError {}
+// impl std::fmt::Display for SourceError {
+//     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+//         match self {
+//             Self::NotFound => write!(f, "File not found"),
+//             Self::InternalError(e) => e.fmt(f),
+//         }
+//     }
+// }
+// impl std::error::Error for SourceError {}
+
+pub fn source_file<P>(path: P) -> MaybeZerror<Zerror>
+where
+    P: AsRef<Path>,
+{
+    let path = path.as_ref();
 
-pub fn source_file(path: impl ToCString) -> MaybeError<SourceError> {
-    let path = path.into_cstr();
-    let result = unsafe { zsys::source(path.as_ptr() as *mut _) };
+    let path_str = path.into_cstr();
+    let result = unsafe { zsys::source(path_str.as_ptr() as *mut _) };
     if result == zsys::source_return_SOURCE_OK {
         Ok(())
     } else {
         Err(match result {
-            zsys::source_return_SOURCE_NOT_FOUND => SourceError::NotFound,
-            zsys::source_return_SOURCE_ERROR => SourceError::InternalError(InternalError),
-            _ => unreachable!(),
+            zsys::source_return_SOURCE_NOT_FOUND => Zerror::FileNotFound(path.to_path_buf()),
+            zsys::source_return_SOURCE_ERROR => Zerror::SourceError(path.to_path_buf()),
+            _ => unreachable!("Unknown source return value"),
         })
     }
 }
+
+/// Changes the current working directory, idk if this works or not
+pub fn cd<P>(path: P) -> MaybeZerror<Zerror>
+where
+    P: AsRef<Path>,
+{
+    let path = path.as_ref();
+    if !path.is_dir() {
+        return Err(Zerror::FileNotFound(path.into()));
+    }
+    let cd_operation = unsafe { zsh_sys::chdir(path.into_cstr().as_ptr() as *mut _) };
+
+    // TODO: Make actually good error handling
+    if cd_operation == 0 {
+        Ok(())
+    } else {
+        Err(Zerror::CdError((path.into(), cd_operation)))
+    }
+}
```

Comment:
We should probably not use `zsh_sys::chdir` because it is a C lib API. I am yet to investigate if zsh has it's own function for this stuff.

In the meantime, `std::env::set_current_dir` probably works